### PR TITLE
Simplify IntersectionSearchSpace.calculate

### PR DESCRIPTION
**Pull request by knshnb**
## Motivation

I simplified the implementation of `IntersectionSearchSpace.calculate` a little.

## Description of the changes

Directly filter the search space dictionary instead of creating a delete list and deleting them one by one.

### CODE REVIEW

**Comment by contramundum53** [comment]
@knshnb This might cause a performance degradation, since the `set` object is created many times throughout the calculation. 
(Originally, `delete_list` is an empty list in most cases.)

**Comment by contramundum53** [comment]
@not522 What do you think?

**Comment by contramundum53** [comment]
I think it is possible to simplify this program using generators without affecting the performance, but the code might be trickier.

**Comment by knshnb** [comment]
Thanks for pointing it out! I confirmed that when `delete_list` is empty, creating a new `dict` object takes longer than just iterating over the existing `dict` in the Google Colab environment.

```python
d = {x: 2 * x for x in range(1000)}
```
```python
%%timeit
for _ in range(10000):
  d1 = {k: v for k, v in d.items() if k >= 0}
```
1 loop, best of 5: 1.02 s per loop
```python
%%timeit
for _ in range(10000):
  delete_list = []
  for k, v in d.items():
    if k < 0:
      delete_list.append(k)
  for k in delete_list:
    del d[k]
```
1 loop, best of 5: 595 ms per loop

**Comment by knshnb** [comment]
This part should probably not be a bottleneck, but I don't have a strong opinion and there might be no need to dare to include this change. What do you think?

**Comment by contramundum53** [comment]
I see, it seems that this change won't make much difference.

File: optuna/samplers/_search_space/intersection.py [hunk 1]
```diff
@@ -83,15 +83,11 @@ def calculate(self, study: Study, ordered_dict: bool = False) -> Dict[str, BaseD
...
-                    delete_list.append(param_name)
-                elif trial.distributions[param_name] != param_distribution:
-                    delete_list.append(param_name)
-
-            for param_name in delete_list:
-                del self._search_space[param_name]
+            self._search_space = {
+                name: distribution
+                for name, distribution in self._search_space.items()
+                if name in trial.distributions and trial.distributions[name] == distribution
```
Comment (not522) [inline optuna/samplers/_search_space/intersection.py:89]:
```suggestion
                if trial.distributions.get(name) == distribution
```

**Review by not522 — COMMENTED** [review]
Basically LGTM. I left one suggestion.

File: optuna/samplers/_search_space/intersection.py (hunk 1 above)
Comment (knshnb) [reply optuna/samplers/_search_space/intersection.py]: Thanks!

**Review by not522 — APPROVED** [review]

**Comment by contramundum53** [comment]
LGTM!

**Pull request merged**